README.md
##[访问http://www.behaviac.com/获取文档，教程，API，FAQ，源码，下载等一切相关资料](http://www.behaviac.com/)

 - behaviac是游戏AI的开发框架组件，也是游戏原型的快速设计工具
 - 支持行为树BT，状态机FSM，HTN等多种范式
 - 方便的编辑，实时和离线调试
 - 支持全平台，适用于客户端和服务器，助力游戏快速迭代开发
 - [官方网站](http://www.behaviac.com/)是文档，教程，API，FAQ，源码，下载等一切的入口
 - 您可以访问[behaviac Q&A](http://bbs.behaviac.com/)获取答案或提出问题
 - 您可以加入我们的QQ群433547396获得即时的帮助或者信息反馈。请大家优先使用[behaviac Q&A](http://bbs.behaviac.com/)。
 - BehaviacSetup*.exe是安装包，内含可执行的编辑器及示例。如需要自行构建，可以访问官网或[github behaviac](https://github.com/TencentOpen/behaviac)
 

---------------------------------------------
##[Visist http://www.behaviac.com/ for documents, tutorials, FAQs and all other information](http://www.behaviac.com/)

 - behaviac is a framework of the game AI development, and it also can be used as a rapid game prototype design tool
 - behaviac supports the behavior tree, finite state machine and hierarchical task network
 - Behaviors can be designed and debugged in the designer, exported and executed by the game
 - The designer can only run on the Windows platforms, The run time library is implemented with C++ and C#, and it supports all major platforms (Windows, Linux, Android, iOS, Unity etc.) and Unity. 
 - The C++ version is suitable for the client and server side. 
 - [Website](http://www.behaviac.com/) for documents, tutorials, API，FAQ，source code, downloads，etc.
 - [behaviac Q&A](http://bbs.behaviac.com/) for answers of your questions.
 - BehaviacSetup*.exe is the setup package with the binary editor and demo executable. You can download/clone the source code from [github behaviac](https://github.com/TencentOpen/behaviac)
 
 

history.txt
2016-10-17  3.5.16
    Fix a bug for selecting the agent type.
    Fix a bug for setting the parameters of the method for the Effector.

2016-09-26  3.5.15
    Fix a bug for checking the Selector node.
    Fix a bug for the type exported path.
    Fix a bug for the string value with space.
    Fix a bug for the default value of the ExportFileCount.

2016-09-20  3.5.14
    Support splitting the generated_behaviors.cpp file.
    Add the BEHAVIAC_CS_ONLY to use the behaviac lib without UnityEngine.
    Fix a bug for the type sbyte.
    Fix a bug for the ref type of the parameter of the customized method.

2016-08-31  3.5.13
    Fix a bug for sending data when the network is disconnected.
    Fix a bug for generating the return type for the customized method.
    Sort the meta types before exporting types.
    export base agent types content when it is not decorated by type attribute
    
2016-08-24  3.5.12
    fix a bug when there is a ResultStatus
    Fix a bug for using the customized types    
    Fix a bug for the StatusResult method
    Equanl renamed to _Object_Equal_
    operator== and operator!= removed for TagObject

2016-08-18  3.5.11
    Fix a bug for saving the Language of the workspace.

2016-08-16  3.5.10
    assignment supports cast
    fix bug when creating an Agent custom variable

2016-08-08  3.5.9
    fix a bug in c# about member field logging
    log the current variables, otherwise, its value is not the latest
    fix a bug when buffer is reallcated
    fix a bug when double is used
    fix compiling errors on linux
    fix a bug for setting the relative path for the generating the types
    add logging of export info

2016-08-01  3.5.8
    fix possible memory leak when subtree
    fix a bug in Vector to setList when returning from the pool
    fix a possible crash in FSM due to negative nextStateId
    fix a possible crash for vector usage
    socket connection restored for mac
    
2016-07-12  3.5.7
    fix a bug for the array member of the struct.
    fix a bug for generating the type files when the language is null.
    fix a bug when a struct has a vector member of a struct
    fix compiling error on iOS

2016-07-07  3.5.6
    fix a bug(file path) in HotReload of unity version
    add prompt messages if not adding generated cs into the project
    fix a bug in ullong
    fix a bug in editor when the same property is used with different instance name
    fix a bug for selecting a parameter with the list type
    ListParam added to restrict item type for List
    fix bug in edtior for VectorAdd, etc.
    fix a bug for the instance of the property

2016-06-30  3.5.5
    refactor Compare and Compute to avoid Expression Lamda for bugs on iOS
    refine referencebehavior to avoid possible memory leak
    fix bug in compute when right side is function 
    FireEvent can't be called when btexec
    transition refined to support transition when success/failure/exit
    add BEHAVIAC_ prefix for macros and add in behaviac namespace 

2016-06-23  3.5.4
    DoneWithinFrame added to replace repeat
    fix bug in DecoratorLoop for cpp/c# format
    ShowProfilingInfo false by default
    fix bugs in DesignerFloat to support double
    fix bugs when a struct has an array member
    fix bugs in socket connect on linux
    fix a bug for the Local Var when debugging
    fix bug when no xml meta file when creating agent

2016-06-14  3.5.3
    don't export behaviac.bb for cs
    property refactored for c#
    opt struct and vector equals
    fix bug in EBTStatus

2016-06-07  3.5.1
    fix bug in profile display
    Fix a bug for generating codes for the C# structs
    destructor commented in C#
    class can be configured as a value type
    Add the generated type ui on the Meta Browser.

2016-05-31  3.5.0
    to handle the case of no meta xml files
    csharp GC free even on xml/bson

2016-05-20  3.4.4
    to include version number for unitypackage 
    enable memory_leak_test for all platforms
    add SetBehaviorNodeLoader
    rapidxml added to namespace of behaviac
    detect dead loop in the fsm updating
    fix compiling warnings in vs2015 64

2016-05-11  3.4.3
    fix cmake xcode _DEBUG
    fix cmake debug config with the wrong BEHAVIAC_RELEASE
    fix bug that property is not updated in the connected designer   

2016-05-05  3.4.2
    fix bug in task due to a ended running sub node
    fix bug when sub tree reenter

2016-04-27  3.4.1
    DecoratorLog, DecoratorAwaysRunning, DecoratorAwaysSuccess, DecoratorAwaysFailure reshown in the node list
    update cmake scripts to handle LTO, esp. for gcc
    fix bug in FireEvent, passing param by reference

2016-04-21  3.4.0
    subtree refactored to support variable/method to allow dynamic tree

2016-04-13  3.3.16
    64 bit supported, esp. in CMakelists.txt fro win

2016-04-06  3.3.15
    Set the ShowNodeId be to true by default.
    Add the GetRunningNodes() method of the class BehaviorTask.
    Set the UTF-8 format for the control doc.

2016-03-31  3.3.14
    Fix a bug for adding or removing agents when updating all agents.
    Show the behaviac::EBTStatus by default.
    Set the type of the id to int16_t.
    Add F1 for the documents of all nodes.

2016-03-23  3.3.13
    Fix a bug for the readonly property as the attachment.

2016-03-22  3.3.12
    Send the platform info of the runtime part.
    Add the QA menu item for the editor.
    Fix some bugs for the editor.

2016-03-14  3.3.11
    define _DEBUG when debug configuration

2016-03-08  3.3.10
    behaviors.dbg.xml is not used any more

2016-03-02  3.3.9
    The C++/C# export path shoule be the same drive as the workspace.
    Change the web link for stats.

2016-02-29  3.3.8
    Add the automatic signature for the setup package.

2016-02-25  3.3.7
    Fix a bug for the parameters of method generation.

2016-02-24  3.3.6
    Fix a bug for resetting the node id when replacing the node.
    Report the username and mac address to the TQOS.

2016-02-23  3.3.5

src/base/md5.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/base/md5.h"
#include "behaviac/base/swapbytehelpers.h"

static uint32_t MD5IMPL_F(uint32_t x, uint32_t y, uint32_t z)
{
    return (x & y) + ((~x)&z);
}
static uint32_t MD5IMPL_G(uint32_t x, uint32_t y, uint32_t z)
{
    return (x & z) | (y & (~z));
}
static uint32_t MD5IMPL_H(uint32_t x, uint32_t y, uint32_t z)
{
    return (x ^ y ^ z);
}
static uint32_t MD5IMPL_I(uint32_t x, uint32_t y, uint32_t z)
{
    return (y ^ (x | (~z)));
}
static uint32_t MD5IMPL_rol(uint32_t x, uint32_t n)
{
    return (x << n) | (x >> (32 - n));
}

#define MD5_ABCD(F,k,s,Ti) MD5_IMPL(a,b,c,d,F,k,s,Ti)
#define MD5_DABC(F,k,s,Ti) MD5_IMPL(d,a,b,c,F,k,s,Ti)
#define MD5_CDAB(F,k,s,Ti) MD5_IMPL(c,d,a,b,F,k,s,Ti)
#define MD5_BCDA(F,k,s,Ti) MD5_IMPL(b,c,d,a,F,k,s,Ti)
#define MD5_IMPL(a,b,c,d,F,k,s,Ti) \
    a = MD5IMPL_rol(a + MD5IMPL_##F(b,c,d) + m_buf.w[k] + Ti, s) + b;

BEHAVIAC_FORCEINLINE /* only one call site */
void CMD5Hash::ProcessBlock()
{
    uint32_t a = m_abcd.w[0];
    uint32_t b = m_abcd.w[1];
    uint32_t c = m_abcd.w[2];
    uint32_t d = m_abcd.w[3];
    // Round 1
    MD5_ABCD(F, 0, 7, 0xD76AA478)  MD5_DABC(F, 1, 12, 0xE8C7B756)
    MD5_CDAB(F, 2, 17, 0x242070DB)  MD5_BCDA(F, 3, 22, 0xC1BDCEEE)
    MD5_ABCD(F, 4, 7, 0xF57C0FAF)  MD5_DABC(F, 5, 12, 0x4787C62A)
    MD5_CDAB(F, 6, 17, 0xA8304613)  MD5_BCDA(F, 7, 22, 0xFD469501)
    MD5_ABCD(F, 8, 7, 0x698098D8)  MD5_DABC(F, 9, 12, 0x8B44F7AF)
    MD5_CDAB(F, 10, 17, 0xFFFF5BB1)  MD5_BCDA(F, 11, 22, 0x895CD7BE)
    MD5_ABCD(F, 12, 7, 0x6B901122)  MD5_DABC(F, 13, 12, 0xFD987193)
    MD5_CDAB(F, 14, 17, 0xA679438E)  MD5_BCDA(F, 15, 22, 0x49B40821)
    // Round 2
    MD5_ABCD(G, 1, 5, 0xF61E2562)  MD5_DABC(G, 6, 9, 0xC040B340)
    MD5_CDAB(G, 11, 14, 0x265E5A51)  MD5_BCDA(G, 0, 20, 0xE9B6C7AA)
    MD5_ABCD(G, 5, 5, 0xD62F105D)  MD5_DABC(G, 10, 9, 0x02441453)
    MD5_CDAB(G, 15, 14, 0xD8A1E681)  MD5_BCDA(G, 4, 20, 0xE7D3FBC8)
    MD5_ABCD(G, 9, 5, 0x21E1CDE6)  MD5_DABC(G, 14, 9, 0xC33707D6)
    MD5_CDAB(G, 3, 14, 0xF4D50D87)  MD5_BCDA(G, 8, 20, 0x455A14ED)
    MD5_ABCD(G, 13, 5, 0xA9E3E905)  MD5_DABC(G, 2, 9, 0xFCEFA3F8)
    MD5_CDAB(G, 7, 14, 0x676F02D9)  MD5_BCDA(G, 12, 20, 0x8D2A4C8A)
    // Round 3
    MD5_ABCD(H, 5, 4, 0xFFFA3942)  MD5_DABC(H, 8, 11, 0x8771F681)
    MD5_CDAB(H, 11, 16, 0x6D9D6122)  MD5_BCDA(H, 14, 23, 0xFDE5380C)
    MD5_ABCD(H, 1, 4, 0xA4BEEA44)  MD5_DABC(H, 4, 11, 0x4BDECFA9)
    MD5_CDAB(H, 7, 16, 0xF6BB4B60)  MD5_BCDA(H, 10, 23, 0xBEBFBC70)
    MD5_ABCD(H, 13, 4, 0x289B7EC6)  MD5_DABC(H, 0, 11, 0xEAA127FA)
    MD5_CDAB(H, 3, 16, 0xD4EF3085)  MD5_BCDA(H, 6, 23, 0x04881D05)
    MD5_ABCD(H, 9, 4, 0xD9D4D039)  MD5_DABC(H, 12, 11, 0xE6DB99E5)
    MD5_CDAB(H, 15, 16, 0x1FA27CF8)  MD5_BCDA(H, 2, 23, 0xC4AC5665)
    // Round 4
    MD5_ABCD(I, 0, 6, 0xF4292244)  MD5_DABC(I, 7, 10, 0x432AFF97)
    MD5_CDAB(I, 14, 15, 0xAB9423A7)  MD5_BCDA(I, 5, 21, 0xFC93A039)
    MD5_ABCD(I, 12, 6, 0x655B59C3)  MD5_DABC(I, 3, 10, 0x8F0CCC92)
    MD5_CDAB(I, 10, 15, 0xFFEFF47D)  MD5_BCDA(I, 1, 21, 0x85845DD1)
    MD5_ABCD(I, 8, 6, 0x6FA87E4F)  MD5_DABC(I, 15, 10, 0xFE2CE6E0)
    MD5_CDAB(I, 6, 15, 0xA3014314)  MD5_BCDA(I, 13, 21, 0x4E0811A1)
    MD5_ABCD(I, 4, 6, 0xF7537E82)  MD5_DABC(I, 11, 10, 0xBD3AF235)
    MD5_CDAB(I, 2, 15, 0x2AD7D2BB)  MD5_BCDA(I, 9, 21, 0xEB86D391)
    m_abcd.w[0] += a;
    m_abcd.w[1] += b;
    m_abcd.w[2] += c;
    m_abcd.w[3] += d;
}

#undef MD5_ABCD
#undef MD5_DABC
#undef MD5_CDAB
#undef MD5_BCDA
#undef MD5_IMPL

//--------------------------------------------------------------------------

void CMD5Hash::Start()
{
    m_msgLength[0] = m_msgLength[1] = 0;
    m_abcd.w[0] = 0x67452301;
    m_abcd.w[1] = 0xEFCDAB89;
    m_abcd.w[2] = 0x98BADCFE;
    m_abcd.w[3] = 0x10325476;
}

//--------------------------------------------------------------------------

void CMD5Hash::append(const uint8_t* data, uint32_t dataSize)
{
    uint32_t bufFill = (m_msgLength[0] >> 3) & 63;

    while (dataSize > 0)
    {
        m_msgLength[0] += 8;
        m_buf.b[bufFill++] = *data++;
        --dataSize;

        if (bufFill >= 64)
        {
            if (m_msgLength[0] == 0)
            {
                m_msgLength[1]++;
            }

#if BEHAVIAC_BIGENDIAN

            for (uint32_t i = 0; i < 16; ++i)
            {
                behaviacSwapByte(m_buf.w[i]);
            }

#endif
            ProcessBlock();
            bufFill = 0;
        }
    }
}

//--------------------------------------------------------------------------

void CMD5Hash::Finish(SMD5Digest& digest)
{
    static const uint8_t padding[64] =
    {
        0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };
    uint32_t count[2];
    count[0] = m_msgLength[0];
    count[1] = m_msgLength[1];
#if BEHAVIAC_BIGENDIAN
    behaviacSwapByte(count[0]);
    behaviacSwapByte(count[1]);
#endif
    uint32_t padLength = 64 - (((m_msgLength[0] >> 3) + 8) & 63);
    append(padding, padLength);
    append(reinterpret_cast<uint8_t*>(count), 8);
    BEHAVIAC_ASSERT((m_msgLength[0] & 63) == 0);
#if BEHAVIAC_BIGENDIAN
    behaviacSwapByte(m_abcd.w[0]);
    behaviacSwapByte(m_abcd.w[1]);
    behaviacSwapByte(m_abcd.w[2]);
    behaviacSwapByte(m_abcd.w[3]);
#endif

    for (uint32_t i = 0; i < 16; ++i)
    {
        digest.m_digest[i] = m_abcd.b[i];
    }

    // *** If you were going to use this hash function for anything
    // *** sensitive, you would want to clear its state here...
    //m_msgLength[0] = m_msgLength[1] = 0;
    //memset(&m_buf, 0, sizeof(m_buf));
    //memset(&m_abcd, 0, sizeof(m_abcd));
}

//-------------------------------------------------------------------------

#ifdef DEBUG_MD5HASH

// this test code is adapted from the appendix of RFC-1321.
static const char* const test[7 * 2] =
{
    "", "d41d8cd98f00b204e9800998ecf8427e",
    "a", "0cc175b9c0f1b6a831c399e269772661",
    "abc", "900150983cd24fb0d6963f7d28e17f72",
    "message digest", "f96b697d7cb7938d525a2f31aaf161d0",
    "abcdefghijklmnopqrstuvwxyz", "c3fcd3d76192e4007dfb496cca67e13b",
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789", "d174ab98d277d9f5a5611c2c9f419d9f",
    "12345678901234567890123456789012345678901234567890123456789012345678901234567890", "57edf4a22be3c955ac49da2e2107b67a"
};

void TestMD5()
{
    int status = 0;

    for (int i = 0; i < sizeof(test) / sizeof(test[0]); i += 2)
    {
        CMD5Hash hash;
        SMD5Digest digest;
        hash.Start();
        uint32_t length = (uint32_t)strlen(test[i]);
        hash.append((const uint8_t*)test[i], length);
        hash.Finish(digest);
        char hex_output[16 * 2 + 1];

        for (int di = 0; di < 16; ++di)
        {
            string_sprintf(hex_output + di * 2, "%02x", digest.m_digest[di]);
        }

        if (strcmp(hex_output, test[i + 1]))
        {
            BEHAVIAC_LOGINFO("MD5 (\"%s\") = %s", test[i], hex_output);
            BEHAVIAC_LOGINFO("**** ERROR, should be: %s\n", test[i + 1]);
            status = 1;
        }
    }

    if (status == 0)
    {
        BEHAVIAC_LOGINFO("md5 self-test completed successfully.");
    }

    return status;
}
#endif  // DEBUG_MD5HASH

//-------------------------------------------------------------------------

//-------------------------------------------------------------------------
